Add HIP for obtaining output from hook jobs and pods (#301)
helm/community: hips/hip-9999.md

---
hip: 9999
title: "New annotations for displaying job output"
authors: [ "Ian Zink <ian@replicated.com>" ]
created: "2023-01-26"
type: "feature"
status: "draft"
---

## Abstract

> This proposes a new annotation to indicate that the output from the job should be displayed to the user.

Reviewer: In this case, I think you mean the output of a hook not a job. A hook may be implemented with a job but it's the hook.

For example, a job that's not a hook would not be linked to this annotation. I think the title needs to reflect any change here.

## Motivation

The primary motivation for this HIP is the ability to display Preflight checks before the main chart installs. This provides vital feedback to the user as to why the chart install is not proceeding.

> Often it is important to verify that the kubernetes cluster you are deploying a helm chart into has certain properties. You might need to know that the cluster is of a certain version to use various APIs. You might need to know that it has ingress available, a certain amount of ephemeral storage, memory, or CPUs available. You might want to validate the the service key they provided was correct or that that database they entered is reachable. Letting a chart deploy and then finding debugging to see why it failed is a poor user experience. These things can all be done with preflight checks enabled by the hook proposed in this HIP.

Author: For the first, I know that Helm does provide the ability to check for api versions, but there might be good reasons you want to declare it in a tool specifically designed to declare and report on the readiness of a cluster for your app. But for the purposes of this HIP I do see where it could be confusing so I'll just take it out as it's not really crucial here anyway.

For the second, while you can check if an ingress api is available, as far as I'm aware there is not a way to tell if there is a controller that will do anything with it -- which is what this is referring to so I'll be more specific.

> Often it is important to verify that the kubernetes cluster you are deploying a helm chart into has certain properties. You might need to know that the cluster is of a certain version to use various APIs. You might need to know that it has ingress available, a certain amount of ephemeral storage, memory, or CPUs available. You might want to validate the the service key they provided was correct or that that database they entered is reachable. Letting a chart deploy and then finding debugging to see why it failed is a poor user experience. These things can all be done with preflight checks enabled by the hook proposed in this HIP.

Author: I'm going to resolve this let me know if you think it still needs clarification.

> In general, allowing chart developers to run jobs and present that feedback directly to the users could also open up additional use cases beyond just the preflight use case that motivated this HIP. I could imagine scenarios where maybe CVE warnings are presented or specific upgrade feedback is presented instead of just helm install failure.

Author: @Bez625 @xavpaice any thoughts on this one? My first thought is it should be pretty similar to how notes are processed, but I don't think I've ever seen chart notes in flux or argo, but now I should go look.

> In general, allowing chart developers to run jobs and present that feedback directly to the users could also open up additional use cases beyond just the preflight use case that motivated this HIP. I could imagine scenarios where maybe CVE warnings are presented or specific upgrade feedback is presented instead of just helm install failure.

Reviewer 2: My motivation for implementing the feature was to get hook output back to the client (NOT specifically the user). This was more to get feedback in the form of logs to any CD controllers executing helm, but it is also useful to see this if a user is calling helm manually. 

My take is that "preflight" checks are a usecase for hooks. Adding output of hooks to the client is a nice feature to add to hooks. If the issue here is a disagreement with the examples can we not just remove them and stick with something more general. 

E.g. "There are many usecases for hooks and it would improve the user experience if it were possible to output the logs of those hooks when the client executes them. The feedback could be direct to a user, part of a CI/CD pipeline or maybe even consumed from the helm library."

Focusing this on preflight checks could be a bit misleading as I'm pretty sure the implementation works for pre and post hooks, so it really is a new feature for hooks rather than specific to pre-install hooks or the preflight usecase.  

@mattfarina @z4ce does this help or have I missed the point?

> In general, allowing chart developers to run jobs and present that feedback directly to the users could also open up additional use cases beyond just the preflight use case that motivated this HIP. I could imagine scenarios where maybe CVE warnings are presented or specific upgrade feedback is presented instead of just helm install failure.

Author: @mattfarina can chime in here some more, but I think the real point of this comment is about how would Flux/Argo be able to make use of the output that helm gathered? 

When an API user gets a Release ( https://pkg.go.dev/helm.sh/helm/v3@v3.12.3/pkg/release#Release ) back, how should they then get the relevant logs?

My first thought was that we should put it in something like Release.Info.HookOutput[], but that object gets stored in the secret that helm manages and putting more data into there doesn't seem desirable.

> In general, allowing chart developers to run jobs and present that feedback directly to the users could also open up additional use cases beyond just the preflight use case that motivated this HIP. I could imagine scenarios where maybe CVE warnings are presented or specific upgrade feedback is presented instead of just helm install failure.

Reviewer 2: > @mattfarina can chime in here some more, but I think the real point of this comment is about how would Flux/Argo be able to make use of the output that helm gathered?

My reading of the comment was that IF you accept the usecase presented (i.e. specifically for preflight checks) then you need to accommodate for that in any way that those preflight checks could be called. My view is that there was never an intention on my part to implement this specifically for preflight checks - the purpose was to give feedback as a convenience for users of the client (which could be called via a CD process that has its own log handling and was the case we wanted it for). 

I think it best @mattfarina add clarity to his comment at this stage as it seems we aren't aligned on the intent of his feedback.  

> When an API user gets a Release ( https://pkg.go.dev/helm.sh/helm/v3@v3.12.3/pkg/release#Release ) back, how should they then get the relevant logs?

I strongly feel they shouldnt. The Release describes a helm deployment and the log output from a job or pod is way too fine a level of detail for that struct. E.g. if a kubernetes Deployment failed with CrashLoopBackoff you wouldnt expect the logs of multiple pods to appear on the Release struct. 

I think if the desire was to implement a preflight checks feature you would opt for something much more opinionated - arguably a first class citizen of the codebase. It would likely not be an extension of hooks, but it's own concept and have well defined constructs for declaring each check and the results format it should conform to. An array of []PreFlightChecks would make more sense on the Release struct. 

Personally, I feel outputting of hook logs is a strong enough usecase on it's own and there is clearly a lot of support for it in the community. Since this PR has been sat around for nearly 2 years I'd like to try and get it merged in. If the delay is simply a disagreement in the Motivation then let's change the motivation to what it was intended for, i.e. a feature addition to hooks that adds a convenience for users of the CLI.

> In general, allowing chart developers to run jobs and present that feedback directly to the users could also open up additional use cases beyond just the preflight use case that motivated this HIP. I could imagine scenarios where maybe CVE warnings are presented or specific upgrade feedback is presented instead of just helm install failure.

Author: @Bez625 is your CD system using the helm CLI directly so that the output from the CLI would meet your needs in your CD system? If so, I think that's a great data point but we still need to think about how SDK users should interact with this.

> In general, allowing chart developers to run jobs and present that feedback directly to the users could also open up additional use cases beyond just the preflight use case that motivated this HIP. I could imagine scenarios where maybe CVE warnings are presented or specific upgrade feedback is presented instead of just helm install failure.

Reviewer 2: > @Bez625 is your CD system using the helm CLI directly so that the output from the CLI would meet your needs in your CD system? 

Correct 

> If so, I think that's a great data point but we still need to think about how SDK users should interact with this.

I really dont think we do. The feature was always intended to give log output to the CLI as a convenience. 

Given we have [more and more](https://github.com/helm/helm/pull/10309#issuecomment-1720782194) support for the change as it was intended can we not just merge that?

> In general, allowing chart developers to run jobs and present that feedback directly to the users could also open up additional use cases beyond just the preflight use case that motivated this HIP. I could imagine scenarios where maybe CVE warnings are presented or specific upgrade feedback is presented instead of just helm install failure.

Author: Looping in @hiddeco from FluxCD.

> In general, allowing chart developers to run jobs and present that feedback directly to the users could also open up additional use cases beyond just the preflight use case that motivated this HIP. I could imagine scenarios where maybe CVE warnings are presented or specific upgrade feedback is presented instead of just helm install failure.

Reviewer 3: What we do within Flux' helm-controller is that we capture the last 5 log lines in a ring buffer (after de-duplication, as certain actions may push the same message over and over while they are waiting for something), both from the Helm action that is being run as Helm's underlying Kubernetes client (while storage instance logs are just being pushed to the trace level of the controller logs). This information is then emitted in a Kubernetes Event in case of a failure, as this historically is about the only time when you want to know any details about the last steps Helm took.

An example Kubernetes Event message:

```
Helm upgrade failed: context deadline exceeded

Last Helm logs:

Patch Deployment "podinfo-helmrepository" in namespace default
waiting for release podinfo-helmrepository resources (created: 0 updated: 2  deleted: 0)
beginning wait for 2 resources with timeout of 5m0s
Deployment is not ready: default/podinfo-helmrepository. 0 out of 1 expected pods are ready
warning: Upgrade "podinfo-helmrepository" failed: context deadline exceeded
```

Reason we already loop in the Kubernetes client logs is because otherwise timeout failures do not tell you much about the reason of failure in case it's e.g. due to a Deployment not reaching a Ready state (as shown above).

On a quick first glance, I would thus say that this HIP would be supported in combination with how Flux works at present out of the box. But we may have to increase the number of log lines we capture, as people tend to do odd things with charts once they are allowed to, and this is screaming arbitrary output possibilities to me.

However, I am also wondering why this would require any annotation at all. Instead of just allowing the user to provide a flag during install and/or upgrade which simply pipes the logs of (any) hooks to stdout if requested? This would seem more useful to me for debugging purposes than giving the chart author control over what's relevant to the user, especially given the wide range of issues you can run into while operating Kubernetes.

> In general, allowing chart developers to run jobs and present that feedback directly to the users could also open up additional use cases beyond just the preflight use case that motivated this HIP. I could imagine scenarios where maybe CVE warnings are presented or specific upgrade feedback is presented instead of just helm install failure.

Reviewer 3: Thinking more about it, the annotation model has a weakness in combination with post-rendering (which could allow you to inject your own annotations if the chart author doesn't), because hooks are excluded from the post-rendering engine: https://github.com/helm/helm/issues/7891

> In general, allowing chart developers to run jobs and present that feedback directly to the users could also open up additional use cases beyond just the preflight use case that motivated this HIP. I could imagine scenarios where maybe CVE warnings are presented or specific upgrade feedback is presented instead of just helm install failure.

Author: Some times chart author's want to indicate that particular hook output is important -- like a preflight check. And indeed, for ISVs, they likely know what output is more important than the end user. 

Good to know it would work with Flux as is. Given that output I'm inclined to resolve this consideration with "CD tools can obtain output from the helm logs"

The other path I was thinking about was providing another item under configuration that could allow callers to specify an `HookOutput io.Writer` which the output would be written to.

> In general, allowing chart developers to run jobs and present that feedback directly to the users could also open up additional use cases beyond just the preflight use case that motivated this HIP. I could imagine scenarios where maybe CVE warnings are presented or specific upgrade feedback is presented instead of just helm install failure.

Reviewer 3: > The other path I was thinking about was providing another item under configuration that could allow callers to specify an `HookOutput io.Writer` which the output would be written to.

This _may_ be better, but would still make per-hook aggregation hard, which is what a perfect SDK would facilitate. But we could then at least separate the succession hook logs from any other noise we are not interested in as long as we are not running into a failure.

> In general, allowing chart developers to run jobs and present that feedback directly to the users could also open up additional use cases beyond just the preflight use case that motivated this HIP. I could imagine scenarios where maybe CVE warnings are presented or specific upgrade feedback is presented instead of just helm install failure.

Author: We could make it a call back. 

```
HookOutput func(string) io.Writer //string is the container name of the output we're writing
```

Is that the kind of aggregation (or really deaggregation) a perfect sdk would have?

> In general, allowing chart developers to run jobs and present that feedback directly to the users could also open up additional use cases beyond just the preflight use case that motivated this HIP. I could imagine scenarios where maybe CVE warnings are presented or specific upgrade feedback is presented instead of just helm install failure.

Reviewer 3: Think that would give you much more to work and sugar with than plain logs or a single stream of data, yes.

> In general, allowing chart developers to run jobs and present that feedback directly to the users could also open up additional use cases beyond just the preflight use case that motivated this HIP. I could imagine scenarios where maybe CVE warnings are presented or specific upgrade feedback is presented instead of just helm install failure.

Reviewer 4: Chiming in here that one huge motivating example, beyond just a preflight check, and IMO one worth adding to the documentation here for the HIP, is the concept of the release phase from e.g. Heroku and Fly.

https://devcenter.heroku.com/articles/release-phase

https://fly.io/docs/reference/configuration/#run-one-off-commands-before-releasing-a-deployment

For many applications and services with their own relational database(s), this release phase is where database migrations are run, and any failure of a migration should stop the pipeline and prevent the new code (which would assume that e.g. new columns had been successfully added) from deploying. https://atlasgo.io/guides/deploying/helm is an example of a system that recommends this be done in pre-install,pre-upgrade hooks.

But, of course, on such a failure you'd want to know why the failure occurred. Right now, that requires finding and introspecting the pod spawned by the hook's job, which isn't necessarily something that the application developer (who's used to looking at logs in Github Actions or Heroku, say) has permission or training to do.

Since, in the case of multiple hooks, they'd run serially in order (I believe this is what is implied by https://helm.sh/docs/topics/charts_hooks/#hooks-and-the-release-lifecycle ?) a single stream could potentially be fine, if there's a predictable structured delimiter, including the hook name, that's emitted to the stream when each new hook begins executing.

If that's right, then in most cases, I think it would be more than sufficient for the CI/CD system to be set up to just echo this stream to standard output, potentially scanning that output - nothing else would be interleaved, because the hooks must execute before anything else happens.

Apologies if I'm getting anything wrong here or misstepping, as I'm new to contributing to Helm - but I'd love to see this land as soon as possible as it would address a huge pain point for us!

> In general, allowing chart developers to run jobs and present that feedback directly to the users could also open up additional use cases beyond just the preflight use case that motivated this HIP. I could imagine scenarios where maybe CVE warnings are presented or specific upgrade feedback is presented instead of just helm install failure.

Reviewer 2: @mattfarina where do we stand with this review comment? To my mind pre-flight checks and DB maintenance are all great usecases for hooks, but there's really no need to list the usecases for hooks in this PR. This is a feature that enhances working with hooks by allowing hook output to be sent to the client. 

I agree with @z4ce comment in the call that we should follow the YAGNI principle and avoid an SDK specific integration until there is a usecase. That's not to say we should block this PR on finding a usecase for the SDK, but we should merge this and allow someone else to contribute an SDK change when they have a need for it. Also, once the CLI output version of the code is merged it may inspire other code changes that integrate it with other usecases, i.e. allow others in the community to iterate on this first usecase. 

Given @hiddeco has confirmed there could be flux integration with the proposed implementation I think we are satisfied that the PR in it's current form can be leveraged by SDK users.

@bpartridge FWIW we also use the hooks to perform DB maintenance tasks prior to releasing a new version of our app.

> In general, allowing chart developers to run jobs and present that feedback directly to the users could also open up additional use cases beyond just the preflight use case that motivated this HIP. I could imagine scenarios where maybe CVE warnings are presented or specific upgrade feedback is presented instead of just helm install failure.

Author: @hiddeco @mattfarina @Bez625 I went ahead and took the liberty of making a PR to support the method of integration I spoke about above. Let me know if this looks right and I'll write it up for the HIP

https://github.com/Bez625/helm/pull/2/commits/1d7566a5a838f1003933da146fdc20cfcce701a3

> In general, allowing chart developers to run jobs and present that feedback directly to the users could also open up additional use cases beyond just the preflight use case that motivated this HIP. I could imagine scenarios where maybe CVE warnings are presented or specific upgrade feedback is presented instead of just helm install failure.

Reviewer 2: > > I agree with @z4ce comment in the call that we should follow the YAGNI principle and avoid an SDK specific integration until there is a usecase.
> 
> There _is_ a use case, as there are existing SDK dependents. As I mentioned, having access to the job name instead of a stream of log lines is _much more valuable_ to any SDK user.

From the previous comments it wasn't clear to me that there was a usecase for the hook output for SDK users - only that they could potentially access them via the reference implementation, but it might not be the easiest to work with. If you can let me know which interfaces would be best to adapt this for the SDK users (or if you agree with @z4ce suggestion) I'm happy to include that in the reference implementation. 

> Cutting corners by applying a YAGNI principle most of the time results in code which isn't flexible enough to be changed at a later stage without introducing breaking changes. 

You are correct - as a general rule if software development practices are applied incorrectly the result will be poor software. In this specific case no one is trying to cut corners. The intention was - and always has been - to facilitate hook output to CLI users. If there is a way we can incorporate this change to work with an SDK specific usecase I'm more than happy to include it. 

What I'm trying to avoid is extending the scope of this by including a lot of changes to support possible SDK integrations that may not ever be used. If there are slight tweaks we can make to the implementation that will allow future integrations much easier to do (e.g. include job names etc) i'm happy to do that.

> Which is also why it works best in combination with "continuous refactoring", a practice which can not be applied easily to software tied to semantic versioning, and has historically not been applied to Helm.

Based on this experience, I agree that the Helm code lifecycle is not a good fit for "continuous refactoring". 

@z4ce I had a quick look at the change and it looks good to me. I'd like to review the implementation in more detail, but let's confirm @hiddeco is happy with the proposed changes on this thread first.

> In general, allowing chart developers to run jobs and present that feedback directly to the users could also open up additional use cases beyond just the preflight use case that motivated this HIP. I could imagine scenarios where maybe CVE warnings are presented or specific upgrade feedback is presented instead of just helm install failure.

Author: I added this section to the HIP https://github.com/helm/community/pull/301/files#diff-b5f8adc4453184fee3729ff0b5387b3e6c11b174019d59f76e50b87bf9ca73e2R44  I think that this resolves this but before I close this thread I'll give it a couple of days. I'd love to hear from @hiddeco that this interface would work well for him.

> In general, allowing chart developers to run jobs and present that feedback directly to the users could also open up additional use cases beyond just the preflight use case that motivated this HIP. I could imagine scenarios where maybe CVE warnings are presented or specific upgrade feedback is presented instead of just helm install failure.

Reviewer 3: After reviewing the code of the PR while taking into account the note around `c.Log`, I think this approach is fine and gives SDK users sufficient things to work with. It's also better than the previous reliance on `log.Writer()`, which would cause the logging to not be configurable at all due to Go preventing any changes to be made to the `std` instance. As noted in https://github.com/Bez625/helm/commit/1d7566a5a838f1003933da146fdc20cfcce701a3#r127777881 however, it does change the behavior of the PR as instead of `os.Stderr` things will be written to `os.Stdout`.

As I has missed this subtle difference while writing https://github.com/helm/community/pull/301#discussion_r1327826950, it would also have meant that _without_ this change we would be doomed due to it not making use of the logger instance.

In general, my advise would still be to organize better means than relying on a hard-coded (albeit configurable) `os.Stdout` in a code base where appropriate loggers appear to be wired through. But in your defense, there is now at least a way for this to be avoided.

## Security implications

Potentially the preflight checks could check for security misconfigurations that could enhance the security of the chart deployment.

## How to teach this

In the first instance, documentation plus the help text for `helm install` would explain the feature.

An example template could be provided in documentation showing how to use this feature with a generic command used in a hook.

A more advanced example showing how to use the new feature with Troubleshoot.sh to provide preflight checks could be linked in the documentation, provided directly in the documentation, or provided on the Troubleshoot.sh documentation site independently.

## Reference implementation

The `safe-install` plugin (link in references) demonstrates what running preflights could look like, but not in the fashion implemented in this HIP.

> 

Reviewer: `safe-install` isn't a reference implementation of the feature. The PR that's been started for this would be a reference implementation. Can we describe what would change in helm at a high level and point to the PR.

---
hip: 9999
title: "New annotations for displaying hook output"
authors: [ "Ian Zink <ian@replicated.com>" ]
created: "2023-01-26"
type: "feature"
status: "draft"
---

## Abstract

This proposes a new annotation to indicate that the output from the hook should be displayed to the user.

## Motivation

> The one motivation for this HIP is the ability to display Preflight checks before the main chart installs. This provides vital feedback to the user as to why the chart install is not proceeding.

Reviewer 3: ```suggestion
The one motivation for this HIP is the ability to display preflight checks before the main chart installs. This provides vital feedback to the user as to why a Helm chart can not be successfully installed.
```

> Often it is important to verify that the kubernetes cluster you are deploying a helm chart into has certain properties. You might need to know that it has an ingress controller available, a certain amount of ephemeral storage, memory, or CPUs available. You might want to validate the the service key they provided was correct or that that database they entered is reachable. Letting a chart deploy and then finding debugging to see why it failed is a poor user experience. These things can all be done with preflight checks enabled by the hook proposed in this HIP.

Reviewer 3: ```suggestion
Often it is important to verify that the Kubernetes cluster you are deploying a Helm chart into has certain properties. You might need to know that it has an Ingress controller, a certain amount of ephemeral storage, memory, or CPUs available. You might want to validate the service key they provided was correct or that that database they entered is reachable. Letting a Helm install and/or upgrade fail, and then make the user debug this to see why it failed is a poor user experience. These things can all be done with checks enabled by the hook proposed in this HIP.
```

Another common motivation for this HIP is to indicate that database migration output running as a hook should be visible to the user in the case of failure.

> In general, allowing chart developers to run jobs and present that feedback directly to the users could also open up additional use cases beyond just the preflight use case that motivated this HIP.  I could imagine scenarios where maybe CVE warnings are presented or specific upgrade feedback is presented instead of just helm install failure.

Reviewer 3: ```suggestion
In general, allowing chart developers to run jobs and present that feedback directly to the users could also open up additional use cases beyond just the checks use case that motivated this HIP.  I could imagine scenarios where maybe CVE warnings are presented or specific upgrade feedback is presented instead of just a Helm install failure.
```

## Rationale

> There are other ways that this could be implemented. For example, we could have a separate preflight hook type. However, this new hook type wouldn't be handled at all by previous versions of helm. With this design, it requires minimial changes to helm and allows for backwards compatibility.

Reviewer 3: ```suggestion
There are other ways that this could be implemented. For example, we could have a separate preflight hook type. However, this new hook type wouldn't be handled at all by previous versions of Helm. With this design, it requires minimal changes to Helm and allows for backwards compatibility.
```

> Another strategy could be for helm to include Troubleshoot.sh as a dependent library, but this could result in too tight of a coupling between the projects and lower overall flexibility and adaptability. 

Reviewer 3: ```suggestion
Another strategy could be for Helm to include Troubleshoot.sh as a dependent library, but this could result in too tight of a coupling between the projects and lower overall flexibility and adaptability. 
```

## Specification

> Templates could include the following annotations on Batch Jobs:

Reviewer 3: ```suggestion
Templates may include the following annotations on Jobs:
```

In addition, https://helm.sh/docs/topics/charts_hooks/#hooks-and-the-release-lifecycle also mentions:

> If the resource is a Job or Pod kind, Helm will wait until it successfully runs to completion.

Which makes me wonder if Pods should not be included as well?

    
```yaml
        "helm.sh/hook": pre-install, pre-upgrade
        "helm.sh/hook-output-log-policy": hook-failed # or hook-succeeded or hook-failed,hook-succeeded
```

> `helm.sh/hook-output-log-policy` would indicate that helm should display the output of the job to the user.

Reviewer 3: ```suggestion
`helm.sh/hook-output-log-policy` would indicate that Helm should display the output of the Job to the user.
```

> Additionally a new user flag should be created `--no-log-output`  that would skip the output of logs.

Reviewer 3: ```suggestion
Additionally, a new user flag should be created `--no-log-output`  that would skip the output of logs.
```

Additionally, there will be a new item added to the action SDK configuration to allow SDK consumers to get the output. 

> By default this output will be written to stdout, but an SDK consumer can overwrite the HookOutputFunc to provide a custom writer.

Reviewer: I don't think the SDK should write to StdOut. StdErr is where logs usually go. Consider a service, like Flux, running this update. How can things be low impact on them? What about writing to `io.Discard` by default. They set a writer if they want to capture the output. This is how they can opt-in to getting that output.

> By default this output will be written to stdout, but an SDK consumer can overwrite the HookOutputFunc to provide a custom writer.

Reviewer 5: ```suggestion
By default this output will be discarded by writing to `io.Discard`, but an SDK consumer can overwrite the HookOutputFunc to provide a custom writer.
```

```go
type Configuration struct {
  ...

> // Called with container name and returns and expects writer that will receive the log output
> 	HookOutputFunc func(namespace string, pod string, container string) io.Writer

Reviewer 3: ```suggestion
	// Called with container name and returns and expects writer that will receive the log output
	HookOutputFunc func(namespace, pod, container string) io.Writer
```

}
```

## Backwards compatibility

> The only backwards compatibility concern would be that scripts parsing `helm install` output would see some additional text in the case of logs being output. The fact that notes already make the output unstructured should mitigate any concern here. Since we already are trusting chart developers to provide output in the form of notes, this is a logical extension of that that allows the developer to provide more dynamic output. 

Reviewer: What if the Helm client wrote them to StdErr instead? This is where logs like this are usually written. You can even see it in default Go logger.

Reviewer 6: the HIP was updated with this

> The only backwards compatibility concern would be that scripts parsing `helm install` output would see some additional text in the case of logs being output. The fact that notes already make the output unstructured should mitigate any concern here. Since we already are trusting chart developers to provide output in the form of notes, this is a logical extension of that that allows the developer to provide more dynamic output. 

Reviewer 5: ```suggestion
The only backwards compatibility concern would be that scripts parsing `helm install` output would see some additional text in the case of logs being output. The fact that notes already make the output unstructured should mitigate any concern here. Since we already are trusting chart developers to provide output in the form of notes, this is a logical extension of that that allows the developer to provide more dynamic output. The logs would be written to `stderr`, further mitigating any breaking changes to scripts using `helm` CLI.
```

Reviewer 5: @z4ce I added a some suggestions in the address Matt's comments. If you are happy with them, feel free to accept them so we can get a re-review of the PR

    app.kubernetes.io/version: {{ .Chart.AppVersion }}
    helm.sh/chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
  annotations:
    "helm.sh/hook": pre-install, pre-upgrade
    "helm.sh/hook-output-log-policy": hook-failed
    "helm.sh/hook-weight": "-5"
    "helm.sh/hook-delete-policy": hook-succeeded, hook-failed

spec:
  backoffLimit: 0
  template:
    metadata:
      name: "{{ .Release.Name }}"
      labels:
        app.kubernetes.io/managed-by: {{ .Release.Service | quote }}
        app.kubernetes.io/instance: {{ .Release.Name | quote }}
        helm.sh/chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
    spec:
      restartPolicy: Never
      volumes:
        - name: preflights
          configMap:
            name: "{{ .Release.Name }}-preflight-config"
      containers:
      - name: post-install-job
        image: "replicated/preflight:latest"
        command: ["preflight", "--interactive=false", "/preflights/preflights.yaml"]
        volumeMounts:
        - name: preflights
          mountPath: /preflights

---
apiVersion: v1
kind: ConfigMap
metadata:
  annotations:
    "helm.sh/hook": pre-install, pre-upgrade
    "helm.sh/hook-weight": "-6"

>     "helm.sh/hook-delete-policy": hook-succeeded, hook-failed

Reviewer 2: I think these should be on the CronJob/Job/Pod resources and not on the config map?

        {{ end }}
      analyzers:
        - clusterVersion:
            outcomes:
              - fail:
                  when: "< 1.16.0"
                  message: The application requires at least Kubernetes 1.16.0, and recommends 1.18.0.
                  uri: https://kubernetes.io
              - warn:
                  when: "< 1.18.0"
                  message: Your cluster meets the minimum version of Kubernetes, but we recommend you update to 1.18.0 or later.
                  uri: https://kubernetes.io
              - pass:
                  message: Your cluster meets the recommended and required versions of Kubernetes.
        {{ if eq .Values.mariadb.enabled false }}
        - mysql:
            checkName: Must be MySQL 8.x or later
            collectorName: mysql
            outcomes:
              - fail:
                  when: connected == false
                  message: Cannot connect to MySQL server
              - fail:
                  when: version < 8.x
                  message: The MySQL server must be at least version 8
              - pass:
                  message: The MySQL server is ready
        {{ end }}
```

> What it should loook when running:

Reviewer 3: ```suggestion
Which should yield the following output to stdout:
```
